shared/data/core/src/main/java/com/ivy/data/repository/RepositoryMemo.kt

```diff
@@ -0,0 +1,121 @@
+package com.ivy.data.repository
+
+import com.ivy.base.threading.DispatchersProvider
+import com.ivy.data.DataObserver
+import com.ivy.data.DataWriteEvent
+import com.ivy.data.DeleteOperation
+import com.ivy.data.model.sync.Syncable
+import com.ivy.data.model.sync.UniqueId
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class RepositoryMemoFactory @Inject constructor(
+    private val dataObserver: DataObserver,
+    private val dispatchers: DispatchersProvider,
+) {
+    fun <T : Syncable<TID>, TID : UniqueId> createMemo(
+        getDataWriteSaveEvent: (List<T>) -> DataWriteEvent,
+        getDateWriteDeleteEvent: (DeleteOperation<TID>) -> DataWriteEvent
+    ): RepositoryMemo<T, TID> = RepositoryMemo(
+        dataObserver = dataObserver,
+        dispatchers = dispatchers,
+        getDataWriteSaveEvent = getDataWriteSaveEvent,
+        getDataWriteDeleteEvent = getDateWriteDeleteEvent,
+    )
+}
+
+class RepositoryMemo<T : Syncable<TID>, TID : UniqueId> internal constructor(
```

Comment:
Hey @Vishwa-Raghavendra wdyt? Haven't added tests, yet but this reduces the code greatly and will prevent memorization errors in the repos